semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala

```diff
@@ -1,86 +1,64 @@
-package scala.meta.internal.metacp
+package scala.meta.internal.javacp
 
 import java.nio.file.Files
 import java.nio.file.Path
-
 import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 import scala.collection.mutable.ListBuffer
-import scala.meta.internal.metacp.asm.JavaTypeSignature._
-import scala.meta.internal.metacp.asm._
+import scala.meta.internal.javacp.asm._
+import scala.meta.internal.metacp._
 import scala.meta.internal.semanticdb3.SymbolInformation.{Kind => k}
 import scala.meta.internal.{semanticdb3 => s}
-import scala.tools.asm.ClassReader
 import scala.tools.asm.tree.ClassNode
 import scala.tools.asm.tree.FieldNode
 import scala.tools.asm.tree.InnerClassNode
 import scala.tools.asm.tree.MethodNode
 import scala.tools.asm.{Opcodes => o}
 
 object Javacp {
-
-  def sinfos(
-      root: Path,
-      file: Path,
-      isVisited: mutable.Set[Path]
-  ): Seq[s.SymbolInformation] = {
-    sinfosFromOuterClass(root, file, 0, Scope.empty, isVisited)
-  }
-
-  def sinfosFromOuterClass(
-      root: Path,
-      file: Path,
-      outerClassAccess: Int,
-      scope: Scope,
-      isVisited: mutable.Set[Path]): Seq[s.SymbolInformation] = {
-    if (isVisited(file)) Nil
-    else {
-      isVisited += file
-      val bytes = Files.readAllBytes(file)
-      val node = parseClassNode(bytes)
-      if (isAnonymousClass(node)) {
-        // Skip anonymous classes like we do for Scala symbols.
-        Nil
-      } else {
-        sinfosClassNode(node, outerClassAccess, scope, root, isVisited)
-      }
+  def parse(classfile: ToplevelClassfile): Option[ToplevelInfos] = {
+    sinfos(classfile, classfile.node, 0, Scope.empty) match {
+      case others :+ toplevel =>
+        Some(ToplevelInfos(classfile, List(toplevel), others.toList))
+      case _ =>
+        None
     }
   }
 
-  def sinfosClassNode(
+  private def sinfos(
+      toplevel: ToplevelClassfile,
       node: ClassNode,
-      outerClassAccess: Int,
-      scope: Scope,
-      root: Path,
-      isVisited: mutable.Set[Path]): Seq[s.SymbolInformation] = {
+      access: Int,
+      scope: Scope): Seq[s.SymbolInformation] = {
 
     val buf = ArrayBuffer.empty[s.SymbolInformation]
     val decls = ListBuffer.empty[String]
 
     def addInfo(
         symbol: String,
         kind: s.SymbolInformation.Kind,
-        access: Int,
```

Comment:
The ordering here was already following the ordering in semanticdb3.proto, `access` is used to set properties which come after kind 😉 The name "access" comes from asm, not semanticdb "accessibility" (which should maybe be called "visibility"?).

Reply:
Uh oh. My bad.